design/object-store-user.md

```diff
@@ -0,0 +1,158 @@
+### Background
+This proposal is a necessary part of another enhancement to [dynamically provision S3 buckets](https://github.com/rook/rook/issues/2480). 
+
+With the merge of [2076](https://github.com/rook/rook/pull/2076) Rook now supports an object user CRD and its generated secret containing the AWS key pairs. To create a new object store user, a developer creates a simple _CephObjectStoreUser_ yaml file defining the object store and the user’s display name. After applying this file a CephObjectStoreUser CRD is created in the object store’s namespace, and the S3 object user is created.
+
+Some issues with this approach are:
+* the developer needs to know the name and namespace of the object store service, both to create the object store user yaml file, and to know, in advance, the name of the generated secret[1]. 
+* the object store user needs to reside in the same namespace as the target object store.
+
+<sup>[1]  It is desirable that the app developer be able to produce the pod yaml before creation of the object store user and its generated secret, and, thus, the secret’s name must be predictable since it is consumed by the pod. Today, the secret name consists of “rook-ceph-object-user-” + _objectStore’s metadata.Name_ + _CephObjectStoreUser’s metadata.Name_, and is assumed to live in the object store’s namespace. This enhancement proposes to remove the object store name from the secret’s name. Since object store user names must be unique within a namespace, a secret name derived from the CephObjectStoreUser’s `metadata.Name` is also unique, thus the object store name is not required. This supports an advantage of using a storage class (see below), namely, that the developer does not need to know the name (or namespace) of the object store.</sup>
+
+## Proposed Enhancement
+1. Allow a storage class, which references an object store service, to be defined in the object user API, but continue to support the existing `store` property.
```

Comment:
I think I need a bit of help still understanding the real benefit of introducing the `StorageClass`.  It sounds like the main reason is so that the user that is creating the `CephObjectStoreUser` doesn't have to know the name/namespace of the object store.  However, they still need to discover the name of the `StorageClass` itself and insert that into their manifest.  That is probably better/easier than having to discover the name of the object store, but it still requires discovering something at runtime in their environment and inserting it into their manifest.  Are there other proposed benefits to adding a `StorageClass`?

Also, we aren't using anything unique to the `StorageClass` type here, right? Since we are merely using it as a reference to the object store?  I'm not necessarily advocating for using something other than a `StorageClass`, I'm just trying to understand if the same thing could be accomplished with a cluster scoped CRD that we create.  For example, `CephObjectStoreRef` with a field for the name/namespace of the object store to use.  That technically could be used instead of `StorageClass` here, because we aren't using `StorageClass` for anything besides a generic property bag, right?

Reply:
> ... However, they still need to discover the name of the StorageClass itself and insert that into their manifest.
Correct. Discovery of storage classes is a general k8s issue. Storage class visibility can be allowed via RBAC permission. The answer is that a storage admin will define storage class listing rules per account/namespace such that the users can see the storage classes allowed by the admin.